iroh/examples/ping.rs

```diff
@@ -0,0 +1,82 @@
+use std::{str::FromStr, sync::Arc};
+
+use futures_lite::future;
+use iroh_net::{endpoint, Endpoint, NodeId};
+use iroh_router::{Protocol, ProtocolHandler};
+
+#[derive(Debug)]
+struct Ping(Arc<PingInner>);
+
+const ALPN: &[u8] = b"ping";
+
+impl Ping {
+    fn new(endpoint: Endpoint) -> Self {
+        Self(Arc::new(PingInner { endpoint }))
+    }
+
+    async fn ping(&self, target: NodeId) -> anyhow::Result<()> {
+        let conn = self.0.endpoint.connect(target, ALPN).await?;
+        let (mut send, mut recv) = conn.open_bi().await?;
+        let request = b"Ping".to_vec();
+        send.write_all(&request).await?;
+        send.finish()?;
+        let response = recv.read_to_end(1024).await?;
+        println!("got response: {:?}", String::from_utf8_lossy(&response));
+        Ok(())
+    }
+}
+
+impl ProtocolHandler for PingInner {
+    fn accept(self: Arc<Self>, conn: iroh_net::endpoint::Connecting) -> future::Boxed<anyhow::Result<()>> {
+        Box::pin(async move {
+            let conn = conn.await?;
+            let (mut send, mut recv) = conn.accept_bi().await?;
+            let request = recv.read_to_end(1024).await?;
+            println!("got request: {:?}", String::from_utf8_lossy(&request));
+            let response = b"Pong".to_vec();
+            send.write_all(&response).await?;
+            send.finish()?;
+            conn.closed().await;
+            Ok(())
+        })
+    }
+}
+
+// todo: figure out a way to derive this?
+impl Protocol for Ping {
+    fn protocol_handler(&self) -> Arc<dyn iroh_router::ProtocolHandler> {
+        self.0.clone()
+    }
+
+    fn from_protocol_handler(handler: Arc<dyn iroh_router::ProtocolHandler>) -> Option<Self> {
+        Some(Self(handler.into_arc_any().downcast::<PingInner>().ok()?))
```

Comment:
yeah..having users write this is not great, but you could auto implement this I think

Reply:
You need to know the intermediate type. If you have that you can do it, but it would be quite some machinery.

```rust
trait ProtocolIfIKnowTheIntermediateType {
  type Inner;
}

impl<T: ProtocolIfIKnowTheIntermediateType> Protocol for T {
  // basically the thing from above
}
```

then

```rust
impl ProtocolIfIKnowTheIntermediateType for Ping {
  type Inner = PingInner
}
```

or a derive macro, which feels a bit like overkill for this one LOC.

Reply:
Ok, how about this to simplify it a bit:

```rust
pub trait Protocol: Sized {
    fn protocol_handler(&self) -> Arc<dyn ProtocolHandler>;
    fn from_protocol_handler(handler: Arc<dyn ProtocolHandler>) -> Option<Self>;
    fn downcast_via<T: Any + Send + Sync>(handler: Arc<dyn ProtocolHandler>) -> Option<Self>
    where
        Self: From<Arc<T>>
    {
        Some(Self::from(handler.into_arc_any().downcast::<T>().ok()?))
    }
}
```

and then:
```rust
impl Protocol for Ping {
    fn protocol_handler(&self) -> Arc<dyn iroh_router::ProtocolHandler> {
        self.0.clone()
    }

    fn from_protocol_handler(handler: Arc<dyn iroh_router::ProtocolHandler>) -> Option<Self> {
        Self::downcast_via::<PingInner>(handler)
    }
}
```

Reply:
If we get rid of the ability to downcast the whole problem/boilerplate goes away...